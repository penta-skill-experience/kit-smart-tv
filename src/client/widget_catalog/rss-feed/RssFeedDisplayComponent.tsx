import * as React from "react";

interface RssFeedDisplayProps {
    url: string;
}

export class RssFeedDisplayComponent extends React.Component<RssFeedDisplayProps, any> {

    render() {
        return <div>
            <p>(RSS feed content)</p>
        </div>;
    }
}