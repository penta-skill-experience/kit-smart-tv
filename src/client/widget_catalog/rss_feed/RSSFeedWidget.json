{
  "URL": "http://feeds.bbci.co.uk/news/rss.xml",
  "REFRESH_RATE": 600000
}
