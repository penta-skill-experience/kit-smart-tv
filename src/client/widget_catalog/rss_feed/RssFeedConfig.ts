import {WidgetConfig} from "../WidgetConfig";

export class RssFeedConfig implements WidgetConfig {
    url: string;

    save(): Object {
        return {
            url: this.url,
        };
    }

    load(raw: Object) {
        if (raw.hasOwnProperty("url")) {
            this.url = raw["url"];
        } else {
            throw new Error(`could not load RSS feed config from raw data: ${raw}`);
        }
    }
}