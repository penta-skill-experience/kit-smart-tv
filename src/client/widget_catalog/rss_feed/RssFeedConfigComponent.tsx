import * as React from "react";
import DialogTitle from "@mui/material/DialogTitle";
import {DialogContent, Grid} from "@mui/material";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";

interface RssFeedConfigProps {
    url: string;
    save: (rawConfig: Object) => void;
}

interface RssFeedConfigState {
    urlText: string;
}

export class RssFeedConfigComponent extends React.Component<RssFeedConfigProps, RssFeedConfigState> {


    constructor(props: Readonly<RssFeedConfigProps> | RssFeedConfigProps) {
        super(props);
        this.state = {
            urlText: this.props.url,
        };
    }

    private saveConfig() {
        this.props.save({
            url: this.state.urlText,
        });
    }

    render() {
        return <div>
            <DialogTitle>RSS Feed Settings</DialogTitle>
            <DialogContent>
                <Grid container spacing={2} direction="row" justifyContent="center" alignItems="center">
                    <Grid item xs={12}>
                        <h1>Current Source URL:</h1>
                    </Grid>
                    <Grid item xs={12}>
                        <h1>Change Source URL:</h1>
                    </Grid>
                    <Grid item xs={12}>
                        <TextField
                            value={this.state.urlText}
                            onChange={(event: React.ChangeEvent<HTMLInputElement>) => this.setState({
                                urlText: event.target.value,
                            })}
                        />
                    </Grid>
                    <Grid item xs={12}>
                        <Button variant="outlined" onClick={() => this.saveConfig()}>
                            Save Changes
                        </Button>
                    </Grid>
                </Grid>
            </DialogContent>
        </div>;
    }
}