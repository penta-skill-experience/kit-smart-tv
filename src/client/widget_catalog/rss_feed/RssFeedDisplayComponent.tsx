import * as React from "react";
import Parser from "rss-parser";
import {DisplayComponent, DisplayComponentProps} from "../../widget/DisplayComponent";
import config from "../../../shared/persistence/persistence.config.json";
import * as RSSFeedWidgetConfig from "./RSSFeedWidget.json";

interface RssFeedDisplayState {
    loaded: boolean;
    loadedRss: boolean;
    rssFeed: any;
    rssFeedFallback: string;
}

export class RssFeedDisplayComponent extends DisplayComponent<RssFeedDisplayState> {

    private intervalHandle;

    constructor(props) {
        super(props);
        this.state = {
            loaded: false,
            loadedRss: false,
            rssFeed: undefined,
            rssFeedFallback: undefined,
        };
    }

    componentDidMount() {
        this.fetchRssFeed();

        this.intervalHandle = setInterval(() => this.fetchRssFeed(), RSSFeedWidgetConfig.REFRESH_RATE)
    }

    componentWillUnmount() {
        clearInterval(this.intervalHandle);
    }

    componentDidUpdate(prevProps: Readonly<DisplayComponentProps>, prevState: Readonly<any>, snapshot?: any) {
        if (this.props.config["url"] !== prevProps.config["url"]) {
            this.fetchRssFeed();
        }
    }

    private fetchRssFeed(): void {
        RssFeedDisplayComponent.fetchRssFeedXmlString(this.props.config["url"])
            .then(xmlString => {
                const parser = new Parser();
                return parser.parseString(xmlString)
            })
            .then(feed => {
                this.setState({
                    loaded: true,
                    loadedRss: true,
                    rssFeed: feed,
                    rssFeedFallback: undefined
                });
            })
            .catch(reason => this.props.error(`Failed to parse RSS feed. Reason: ${reason}`));
    }

    private static fetchRssFeedXmlString(url: string): Promise<string> {
        const headers = new Headers();
        headers.append("Content-Type", "application/json");
        const requestOptions = {
            method: 'PUT',
            headers: headers,
            body: JSON.stringify({
                url: url
            })
        };
        return fetch(`${config.DOMAIN}/kvv`, requestOptions)
            .then(resp => resp.text());
    }

    render() {
        if (this.state.loaded) {
            if (this.state.loadedRss) {
                return <div className={"font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"}>
                    <h1>RSS Feed</h1>
                    {this.state.rssFeed.items.map((item, i) =>
                        <div key={i}>
                            <h1>{item.title}</h1>
                            <div dangerouslySetInnerHTML={{__html: item.content}}/>
                            <br/>
                        </div>)}
                </div>;
            } else {
                return <div className={"font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"} dangerouslySetInnerHTML={{__html: this.state.rssFeedFallback}}/>;
            }
        } else {
            return <div className={"font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl 8xl:text-4xl"}>loading...</div>;
        }
    }
}