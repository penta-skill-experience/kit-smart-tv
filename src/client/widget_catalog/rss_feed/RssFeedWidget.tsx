import * as React from "react";
import {Widget} from "../../widget/Widget";
import {RssFeedDisplayComponent} from "./RssFeedDisplayComponent";
import {RssFeedConfigComponent} from "./RssFeedConfigComponent";
import {ConfigComponent} from "../../widget/ConfigComponent";
import {DisplayComponent} from "../../widget/DisplayComponent";
import * as RSSFeedWidgetConfig from "./RSSFeedWidget.json";

export class RssFeedWidget implements Widget {

    getDisplayComponentClass(): typeof DisplayComponent {
        return RssFeedDisplayComponent;
    }

    getConfigComponentClass(): typeof ConfigComponent {
        return RssFeedConfigComponent;
    }

    getTitle(): string {
        return "RSS Feed";
    }

    isConfigurable(): boolean {
        return true;
    }

    getDefaultRawConfig(): Object {
        return {
            url: RSSFeedWidgetConfig.URL,
        }
    }
}