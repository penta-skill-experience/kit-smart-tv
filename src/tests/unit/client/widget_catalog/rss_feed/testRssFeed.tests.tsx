/**
 * @jest-environment jsdom
 */
import React from "react";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {configure, shallow} from "enzyme";
import toJson from "enzyme-to-json";
import {RssFeedDisplayComponent} from "../../../../../client/widget_catalog/rss_feed/RssFeedDisplayComponent";
import * as RSSFeedWidgetConfig from "../../../../../client/widget_catalog/rss_feed/RSSFeedWidget.json";
import fetchMock from "jest-fetch-mock";
fetchMock.enableMocks();

const responseMock = jest.spyOn(Response.prototype, "json");

global.Headers = class {
    constructor() {};
    append(var1 : string, var2 : string) {};
} as jest.Mock;
configure({adapter: new Adapter()});

describe("RSS-Feed Snapshots", () => {

    test("RSS-Feed Snapshot parse successful", async () => {
        fetchMock.mockResponse("\n<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n" +
            "<rss version=\"2.0\">\n" +
            "\n" +
            "<channel>\n" +
            "  <title>W3Schools Home Page</title>\n" +
            "  <link>https://www.w3schools.com</link>\n" +
            "  <description>Free web building tutorials</description>\n" +
            "  <item>\n" +
            "    <title>RSS Tutorial</title>\n" +
            "    <link>https://www.w3schools.com/xml/xml_rss.asp</link>\n" +
            "    <description>New RSS tutorial on W3Schools</description>\n" +
            "  </item>\n" +
            "  <item>\n" +
            "    <title>XML Tutorial</title>\n" +
            "    <link>https://www.w3schools.com/xml</link>\n" +
            "    <description>New XML tutorial on W3Schools</description>\n" +
            "  </item>\n" +
            "</channel>\n" +
            "\n" +
            "</rss> ");
        const wrapper = shallow(<RssFeedDisplayComponent error={(msg => {})} config={RSSFeedWidgetConfig.URL} specialBoldFontColor={"ForestGreen"} specialSubtleFontColor={"DarkOrange"}/>);
        await new Promise(process.nextTick);
        expect(toJson(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    test("RSS-Feed Snapshot parse successful", async () => {
        fetchMock.mockResponse("not an rss feed");
        const wrapper = shallow(<RssFeedDisplayComponent error={(msg => {})} config={RSSFeedWidgetConfig.URL} specialBoldFontColor={"ForestGreen"} specialSubtleFontColor={"DarkOrange"}/>);
        await new Promise(process.nextTick);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test("Rss-Feed Snapshot fallback", async () => {
        fetchMock.mockResponse("\n<?xml version=\"1.0\" encoding=\"UTF-8\" ?>\n" +
            "<rss version=\"2.0\">\n" +
            "\n" +
            "<channel>\n" +
            "  <title>W3Schools Home Page</title>\n" +
            "  <link>https://www.w3schools.com</link>\n" +
            "  <description>Free web building tutorials</description>\n" +
            "  <item>\n" +
            "    <title>RSS Tutorial</title>\n" +
            "    <link>https://www.w3schools.com/xml/xml_rss.asp</link>\n" +
            "    <description>New RSS tutorial on W3Schools</description>\n" +
            "  </item>\n" +
            "  <item>\n" +
            "    <title>XML Tutorial</title>\n" +
            "    <link>https://www.w3schools.com/xml</link>\n" +
            "    <description>New XML tutorial on W3Schools</description>\n" +
            "  </item>\n" +
            "</channel>\n" +
            "\n" +
            "</rss> ");
        const wrapper = shallow(<RssFeedDisplayComponent error={(msg => {})} config={RSSFeedWidgetConfig.URL} specialBoldFontColor={"ForestGreen"} specialSubtleFontColor={"DarkOrange"}/>);
        await new Promise(process.nextTick);
        wrapper.setState({
            loaded: true,
            loadedRss: false
        });
        await new Promise(process.nextTick);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });
});