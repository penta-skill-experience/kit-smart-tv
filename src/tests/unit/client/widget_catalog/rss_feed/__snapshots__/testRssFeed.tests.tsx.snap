// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSS-Feed Snapshots RSS-Feed Config without content 1`] = `
<div>
  <ForwardRef(DialogTitle)>
    RSS Feed Settings
  </ForwardRef(DialogTitle)>
  <ForwardRef(DialogContent)>
    <br />
    <ForwardRef(TextField)
      label="RSS Feed URL"
      onChange={[Function]}
    />
  </ForwardRef(DialogContent)>
</div>
`;

exports[`RSS-Feed Snapshots RSS-Feed Snapshot parse successful 1`] = `
<div
  className="font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"
>
  <h1>
    RSS Feed
  </h1>
  <div
    key="0"
  >
    <h1>
      RSS Tutorial
    </h1>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "New RSS tutorial on W3Schools",
        }
      }
    />
    <br />
  </div>
  <div
    key="1"
  >
    <h1>
      XML Tutorial
    </h1>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "New XML tutorial on W3Schools",
        }
      }
    />
    <br />
  </div>
</div>
`;

exports[`RSS-Feed Snapshots RSS-Feed Snapshot parse successful 2`] = `
<div
  className="font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl 8xl:text-4xl"
>
  loading...
</div>
`;

exports[`RSS-Feed Snapshots Rss-Feed Snapshot fallback 1`] = `
<div
  className="font-light leading-normal sm:text-xs lg:text-base xl:text-base 2xl:text-xl 4xl:text-2xl sm:text-left 8xl:text-4xl"
  dangerouslySetInnerHTML={
    Object {
      "__html": undefined,
    }
  }
/>
`;
